### PR TITLE
wip: backend services

### PR DESCRIPTION
**Pull request by craigatgoogle**
Includes evandbrown's original commit: wip: backend services

Added int tests for scheme agnostic backend feature, and added logic to report error when ambiguous backend name is encountered.

### CODE REVIEW

File: src/bosh-google-cpi/google/backendservice_service/google_backendservice_service.go [hunk 1]
```diff
@@ -128,26 +122,39 @@ func (i GoogleBackendServiceService) RemoveInstance(vmLink string) error {
...
-			return BackendService{}, false, nil
+
+	// Search for a matching backend service amongst an aggregated list containing both global and regional items
+	aggregatedBackendServices, err := i.computeService.BackendServices.AggregatedList(i.project).Do()
+	// TODO(craigatgoogle): Employ server-side name filtering once the API filter bug is fixed, https://b.corp.google.com/issues/80238913
+	if err == nil {
+		var backendService *compute.BackendService
+		for _, scopedList := range aggregatedBackendServices.Items {
+			for _, bs := range scopedList.BackendServices {
+				if bs.Name == id && (bs.Region == "" || strings.Contains(bs.Region, region)) {
```
Comment (evandbrown) [inline src/bosh-google-cpi/google/backendservice_service/google_backendservice_service.go]: If we continue to increment region names (there's an eu-west4 today), `strings.Contains` would find `eu-west1` within `eu-west11`. There's a helper func somewhere around here that will extract a short region name from a fully-qualified one. Maybe worth looking into that?

File: src/bosh-google-cpi/google/backendservice_service/google_backendservice_service.go [hunk 2]
```diff
@@ -128,26 +122,39 @@ func (i GoogleBackendServiceService) RemoveInstance(vmLink string) error {
...
-		backendServiceItem, err = i.computeService.BackendServices.Get(i.project, id).Do()
-	} else {
-		backendServiceItem, err = i.computeService.RegionBackendServices.Get(i.project, region, id).Do()
-	}
-	if err != nil {
-		if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == 404 {
-			return BackendService{}, false, nil
+
+	// Search for a matching backend service amongst an aggregated list containing both global and regional items
+	aggregatedBackendServices, err := i.computeService.BackendServices.AggregatedList(i.project).Do()
```
Comment (evandbrown) [inline src/bosh-google-cpi/google/backendservice_service/google_backendservice_service.go]: Nice!

File: src/bosh-google-cpi/integration/vm_test.go [hunk 3]
```diff
@@ -855,4 +855,136 @@ var _ = Describe("VM", func() {
 		Expect(justInstances(ig)).ToNot(ContainElement(ContainSubstring(vmCID)))
 	})
 
+	It("executes the VM lifecycle with a backend service, without scheme", func() {
```
Comment (evandbrown) [inline src/bosh-google-cpi/integration/vm_test.go]: Do we still have coverage for a backend service with scheme?

Comment (craigatgoogle) [reply src/bosh-google-cpi/integration/vm_test.go]: Yes, the test starting on line 805 covers that case.

Comment (craigatgoogle) [reply src/bosh-google-cpi/integration/vm_test.go]: 753 also provides coverage.

File: src/bosh-google-cpi/google/backendservice_service/google_backendservice_service.go (hunk 1 above)
Comment (craigatgoogle) [reply src/bosh-google-cpi/google/backendservice_service/google_backendservice_service.go]: Done.

File: src/bosh-google-cpi/util/util.go [hunk 4]
```diff
@@ -30,8 +29,10 @@ func RegionFromZone(zone string) string {
 	return ""
 }
 
+var zoneRe = regexp.MustCompile("/zones/([a-zA-Z1-9-]+)/?")
```
Comment (craigatgoogle) [inline src/bosh-google-cpi/util/util.go]: The linter was complaining about the unnecessary types in the declaration statement, so I cleaned them up. Also moved  each closer to its function of concern for clarity.

**Pull request merged**